storm-core/src/clj/backtype/storm/daemon/nimbus.clj

```diff
@@ -346,53 +400,104 @@
-   (FileUtils/writeByteArrayToFile (File. (master-stormcode-path stormroot)) (Utils/serialize topology))
-   (FileUtils/writeByteArrayToFile (File. (master-stormconf-path stormroot)) (Utils/toCompressedJsonConf storm-conf))
-   (if (:code-distributor nimbus) (.upload (:code-distributor nimbus) stormroot storm-id))
-   ))
+  (let [subject (get-subject)
+        storm-cluster-state (:storm-cluster-state nimbus)
+        blob-store (:blob-store nimbus)
+        jar-key (master-stormjar-key storm-id)
+        code-key (master-stormcode-key storm-id)
+        conf-key (master-stormconf-key storm-id)
+        nimbus-host-port-info (:nimbus-host-port-info nimbus)]
+    (if tmp-jar-location ;;in local mode there is no jar
+      (do
+        (.createBlob blob-store jar-key (FileInputStream. tmp-jar-location) (SettableBlobMeta. BlobStoreAclHandler/DEFAULT) subject)
+        (if (instance? LocalFsBlobStore blob-store)
+          (.setup-blobstore! storm-cluster-state jar-key nimbus-host-port-info (get-metadata-version blob-store jar-key subject)))))
+    (.createBlob blob-store conf-key (Utils/toCompressedJsonConf storm-conf) (SettableBlobMeta. BlobStoreAclHandler/DEFAULT) subject)
+    (if (instance? LocalFsBlobStore blob-store)
+      (.setup-blobstore! storm-cluster-state conf-key nimbus-host-port-info (get-metadata-version blob-store conf-key subject)))
+    (.createBlob blob-store code-key (Utils/serialize topology) (SettableBlobMeta. BlobStoreAclHandler/DEFAULT) subject)
+    (if (instance? LocalFsBlobStore blob-store)
+      (.setup-blobstore! storm-cluster-state code-key nimbus-host-port-info (get-metadata-version blob-store code-key subject)))))
+
+(defn- read-storm-topology [storm-id blob-store]
+  (Utils/deserialize
+    (.readBlob blob-store (master-stormcode-key storm-id) (get-subject)) StormTopology))
+
+(defn- get-blob-replication-count [blob-key nimbus]
+  (if (:blob-store nimbus)
+          (-> (:blob-store nimbus)
+            (.getBlobReplication  blob-key get-nimbus-subject)
+            (.get_replication))))
 
 (defn- wait-for-desired-code-replication [nimbus conf storm-id]
   (let [min-replication-count (conf TOPOLOGY-MIN-REPLICATION-COUNT)
         max-replication-wait-time (conf TOPOLOGY-MAX-REPLICATION-WAIT-TIME-SEC)
-        total-wait-time (atom 0)
-        current-replication-count (atom (if (:code-distributor nimbus) (.getReplicationCount (:code-distributor nimbus) storm-id) 0))]
-  (if (:code-distributor nimbus)
-    (while (and (> min-replication-count @current-replication-count)
-             (or (= -1 max-replication-wait-time)
-               (< @total-wait-time max-replication-wait-time)))
+        current-replication-count-jar (if (not (local-mode? conf)) (atom
+                                                                     (get-blob-replication-count (master-stormjar-key storm-id) nimbus))
+                                                                     (atom min-replication-count))
```

Comment:
Conditional here is hard to read the way it is formatted:

Maybe something like this?
```Clojure
jar-replication-count (get-blob-replication-count
                         (master-stormjar-key storm-id)
                         nimbus)
current-replication-count-jar (if (not (local-mode? conf))
                                  (atom jar-replication-count)
                                  (atom min-replication-count))
```